ui/src/components/RemoteAccess/PassControlDialog.jsx

```diff
@@ -11,41 +12,69 @@ function PassControlDialog() {
     state.remoteAccess.observers.find((o) => o.requestsControl),
   );
 
-  function handleSubmit(evt) {
-    evt.preventDefault();
-    const formData = new FormData(evt.target);
+  const showModal = inControl && !!requestingObs;
 
-    dispatch(
-      respondToControlRequest(
-        evt.nativeEvent.submitter.name === 'allow',
-        formData.get('message'),
-      ),
-    );
+  const {
+    register,
+    formState,
+    handleSubmit: makeOnSubmit,
+    setError,
+    reset,
+  } = useForm({ defaultValues: { message: 'Here you go!' } });
+
+  const { isDirty, isSubmitted, errors } = formState;
+
+  async function handleSubmit(data, evt) {
+    const isAllow = evt.nativeEvent.submitter.name === 'allow';
+
+    if (!isAllow && !isDirty) {
+      setError(
+        'message',
+        { message: "Please explain why you're denying this request" },
+        { shouldFocus: true },
+      );
+      return;
+    }
+
+    await dispatch(respondToControlRequest(isAllow, data.message));
   }
 
+  useLayoutEffect(() => {
```

Comment:
The component is always rendered, even when hidden (so it can be animated out, I guess), so we must ensure that the form's state is properly reset if the modal is shown multiple times. React Hook Form takes care of resetting on submit, but this effect takes care of the case where the modal closes because the requester cancelled their request.

`useLayoutEffect` was supposed to reset the form before React committed the changes to the DOM, but after double checking, I realise it is no longer needed. I also added a condition to reset the state only when the modal closes.

Reply:
I see, tricky :)